# TASK 1 Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

In [1]:
# Import libraries 
import pandas as pd
import numpy as np
import os

In [2]:
# Create a path for importing

path = r'C:\Users\peter\Documents\04-2024 Instacart Basket Analysis'

In [3]:
# Import and define dataframe including newly derived columns from the previous exercise
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined_merged_new_variables_derived.pkl'))

In [4]:
# Check dimensions of imported dataframe for consistency
ords_prods_merge.shape

(32404859, 20)

# TASK 2 In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [5]:
ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

# TASK 3 Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

# Compared to the results from the subset, the results from the entire dataframe includes ALL 'department_id' values
# Additionally, the mean values for the department ids present in the subset have all changed.
# Some have increased while others decreased
# For example, in the subset department id 11 had a mean value of ' 11.294069', while in the entire dataframe it has a mean value of '15.694469'

# TASK 4 Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [6]:
# These operations were already performed on the dataframe during the exercise (max_order, loyalty_flag)
ords_prods_merge.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      7.0         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   

   product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0          Soda        77              7     9.0   both  Mid-range product   
1          Soda        77              7     9.0   both  Mid-range product   
2          Soda        77              7     9.0   both  Mid-range product   
3          Soda        77              7     9.0   both  Mid-range product   
4          Soda        77              7     9.0   both  Mid-range product   
5          Soda        77              7     9.0   both  Mid-range product   
6          Soda        77              7     9.0   both  Mid-range product   
7          Soda        77              7     9.0   both  Mid-range product   
8          Soda        77              7     9.0   both  Mid-range product   
9          Soda        77              7     9.0   both  Mid-range produ

In [7]:
# Check loyalty flags
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [8]:
# Check Loyalty flag in dataframe
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(20)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10

# Task 5 The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [9]:
# Use groupby and agg() to determine basic statisitcs of spending habits by loyalty group
ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

# Loyal customer spend the least , new customer spend the most , and regular customer falls in between. The average price of a product purchased by a new customer is about $3 more than that of a loyal customer. This is seen by observing the mean.

# TASK 6 The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [10]:
# Create a new column for the average spending of a customer
ords_prods_merge['average_spend'] = ords_prods_merge.groupby(['user_id'])['prices'].transform('mean')

In [11]:
# Check the new column
ords_prods_merge.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      7.0         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   

   product_name  ...  department_id  prices  _merge    price_range_loc  \
0          Soda  ...              7     9.0    both  Mid-range product   
1          Soda  ...              7     9.0    both  Mid-range product   
2          Soda  ...              7     9.0    both  Mid-range product   
3          Soda  ...              7     9.0    both  Mid-range product   
4          Soda  ...              7     9.0    both  Mid-range product   
5          Soda  ...              7     9.0    both  Mid-range product   
6          Soda  ...              7     9.0    both  Mid-range product   
7          Soda  ...              7     9.0    both  Mid-range product   
8          Soda  ...              7     9.0    both  Mid-range product   
9          Soda  ...              7     9.0    both  Mid-range product   
10         Soda  ...              7   

In [12]:
# View relevant columns 
ords_prods_merge[['user_id', 'average_spend']].head(20)

user_id  average_spend
0         1       6.367797
1         1       6.367797
2         1       6.367797
3         1       6.367797
4         1       6.367797
5         1       6.367797
6         1       6.367797
7         1       6.367797
8         1       6.367797
9         1       6.367797
10       15       3.980556
11       15       3.980556
12       15       3.980556
13       15       3.980556
14       15       3.980556
15       19       8.372059
16       19       8.372059
17       19       8.372059
18       21       7.012683
19       31       8.137793

In [13]:
# Assign spender tag to each customer
ords_prods_merge.loc[ords_prods_merge['average_spend']<10,'spender_flag']='Low_spender'

In [14]:
ords_prods_merge.loc[ords_prods_merge['average_spend']>= 10, 'spender_flag'] = 'High_spender'

In [15]:
# Check spender flags
ords_prods_merge['spender_flag'].value_counts(dropna = False)

spender_flag
Low_spender     31770614
High_spender      634245
Name: count, dtype: int64

In [16]:
ords_prods_merge[['user_id', 'average_spend', 'spender_flag', 'prices']].head(60)

user_id  average_spend spender_flag  prices
0         1       6.367797  Low_spender     9.0
1         1       6.367797  Low_spender     9.0
2         1       6.367797  Low_spender     9.0
3         1       6.367797  Low_spender     9.0
4         1       6.367797  Low_spender     9.0
5         1       6.367797  Low_spender     9.0
6         1       6.367797  Low_spender     9.0
7         1       6.367797  Low_spender     9.0
8         1       6.367797  Low_spender     9.0
9         1       6.367797  Low_spender     9.0
10       15       3.980556  Low_spender     9.0
11       15       3.980556  Low_spender     9.0
12       15       3.980556  Low_spender     9.0
13       15       3.980556  Low_spender     9.0
14       15       3.980556  Low_spender     9.0
15       19       8.372059  Low_spender     9.0
16       19       8.372059  Low_spender     9.0
17       19       8.372059  Low_spender     9.0
18       21       7.012683  Low_spender     9.0
19       31       8.137793  Low_spender     9.0
20       31       8.137793  Low_spender     9.0
21       43       7.444898  Low_spender     9.0
22       43       7.444898  Low_spender     9.0
23       52       6.356287  Low_spender     9.0
24       52       6.356287  Low_spender     9.0
25       52       6.356287  Low_spender     9.0
26       52       6.356287  Low_spender     9.0
27       52       6.356287  Low_spender     9.0
28       52       6.356287  Low_spender     9.0
29       52       6.356287  Low_spender     9.0
30       52       6.356287  Low_spender     9.0
31       52       6.356287  Low_spender     9.0
32       52       6.356287  Low_spender     9.0
33       52       6.356287  Low_spender     9.0
34       52       6.356287  Low_spender     9.0
35       52       6.356287  Low_spender     9.0
36       52       6.356287  Low_spender     9.0
37       67       7.656790  Low_spender     9.0
38       67       7.656790  Low_spender     9.0
39       67       7.656790  Low_spender     9.0
40       67       7.656790  Low_spender     9.0
41       67       7.656790  Low_spender     9.0
42       67       7.656790  Low_spender     9.0
43       67       7.656790  Low_spender     9.0
44       67       7.656790  Low_spender     9.0
45       67       7.656790  Low_spender     9.0
46       67       7.656790  Low_spender     9.0
47       67       7.656790  Low_spender     9.0
48       67       7.656790  Low_spender     9.0
49       67       7.656790  Low_spender     9.0
50       67       7.656790  Low_spender     9.0
51       67       7.656790  Low_spender     9.0
52       67       7.656790  Low_spender     9.0
53       67       7.656790  Low_spender     9.0
54       67       7.656790  Low_spender     9.0
55       67       7.656790  Low_spender     9.0
56       81       8.165385  Low_spender     9.0
57       81       8.165385  Low_spender     9.0
58       82       6.441221  Low_spender     9.0
59       82       6.441221  Low_spender     9.0

# TASK 7 In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer."

In [19]:
# Create a column for frequent users
ords_prods_merge['Customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

C:\Users\peter\AppData\Local\Temp\ipykernel_9504\4137169599.py:2: FutureWarning: The provided callable <function median at 0x000002A9F65C8D60> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['Customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [20]:
# Cheaking for new column
ords_prods_merge.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      7.0         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   
15                     6.0         196                  3          0   
16                     8.0         196                 15          1   
17                     6.0         196                  1          1   
18                     6.0         196                  2          0   
19                     7.0         196                 17          0   

   product_name  ...  _merge    price_range_loc     busiest_day  \
0          Soda  ...    both  Mid-range product  Regularly busy   
1          Soda  ...    both  Mid-range product  Regularly busy   
2          Soda  ...    both  Mid-range product  Regularly busy   
3          Soda  ...    both  Mid-range product      Least busy   
4          Soda  ...    both  Mid-range product      Least busy   
5          Soda  ...    both  Mid-range product  Regularly busy   
6          Soda  ...    both  Mid-range product  Regularly busy   
7          Soda  ...    both  Mid-range product  Regularly busy   
8          Soda  ...    both  Mid-range product  Regularly busy   
9          Soda  ...    both  Mid-range product      Least busy   
10         Soda  ...    both  Mid-range product  Regularly busy   
11         Soda  ...    both  Mid-range product 

In [21]:
# Creating flag on order frequency
ords_prods_merge.loc[ords_prods_merge['Customer_frequency'] > 20, 'frequency_flag'] = 'Non_frequent_customer'

In [22]:
ords_prods_merge.loc[(ords_prods_merge['Customer_frequency']>10) & (ords_prods_merge['Customer_frequency']<=20),
                     'frequency_flag'] = 'Regular customer'

In [23]:
ords_prods_merge.loc[ords_prods_merge['Customer_frequency'] <= 10, 'frequency_flag'] = 'Frequent Customer'

In [24]:
# Check frequency flags
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent Customer        22790523
Regular customer          6927608
Non_frequent_customer     2686728
Name: count, dtype: int64

In [25]:
#Checking for columns
ords_prods_merge[['user_id', 'Customer_frequency', 'frequency_flag']].head(60)

user_id  Customer_frequency         frequency_flag
0         1                20.0       Regular customer
1         1                20.0       Regular customer
2         1                20.0       Regular customer
3         1                20.0       Regular customer
4         1                20.0       Regular customer
5         1                20.0       Regular customer
6         1                20.0       Regular customer
7         1                20.0       Regular customer
8         1                20.0       Regular customer
9         1                20.0       Regular customer
10       15                 8.0      Frequent Customer
11       15                 8.0      Frequent Customer
12       15                 8.0      Frequent Customer
13       15                 8.0      Frequent Customer
14       15                 8.0      Frequent Customer
15       19                 7.0      Frequent Customer
16       19                 7.0      Frequent Customer
17       19                 7.0      Frequent Customer
18       21                 7.0      Frequent Customer
19       31                 6.0      Frequent Customer
20       31                 6.0      Frequent Customer
21       43                10.0      Frequent Customer
22       43                10.0      Frequent Customer
23       52                 8.0      Frequent Customer
24       52                 8.0      Frequent Customer
25       52                 8.0      Frequent Customer
26       52                 8.0      Frequent Customer
27       52                 8.0      Frequent Customer
28       52                 8.0      Frequent Customer
29       52                 8.0      Frequent Customer
30       52                 8.0      Frequent Customer
31       52                 8.0      Frequent Customer
32       52                 8.0      Frequent Customer
33       52                 8.0      Frequent Customer
34       52                 8.0      Frequent Customer
35       52                 8.0      Frequent Customer
36       52                 8.0      Frequent Customer
37       67                 7.0      Frequent Customer
38       67                 7.0      Frequent Customer
39       67                 7.0      Frequent Customer
40       67                 7.0      Frequent Customer
41       67                 7.0      Frequent Customer
42       67                 7.0      Frequent Customer
43       67                 7.0      Frequent Customer
44       67                 7.0      Frequent Customer
45       67                 7.0      Frequent Customer
46       67                 7.0      Frequent Customer
47       67                 7.0      Frequent Customer
48       67                 7.0      Frequent Customer
49       67                 7.0      Frequent Customer
50       67                 7.0      Frequent Customer
51       67                 7.0      Frequent Customer
52       67                 7.0      Frequent Customer
53       67                 7.0      Frequent Customer
54       67                 7.0      Frequent Customer
55       67                 7.0      Frequent Customer
56       81                30.0  Non_frequent_customer
57       81                30.0  Non_frequent_customer
58       82                11.0       Regular customer
59       82                11.0       Regular customer

# TASK 9 export as pickle

In [26]:
# Export the updated dataframe as a pickle file
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined_merged_new_variables_derived_aggregated.pkl'))

In [27]:
# Check shape
ords_prods_merge.shape

(32404859, 24)